apollo-router-core/src/query_planner/mod.rs

```diff
@@ -253,66 +270,129 @@ impl PlanNode {
+        operation: String,
+    }
+
+    struct Variables {
+        variables: Map<String, Value>,
+        paths: Vec<Path>,
+    }
+
+    impl Variables {
+        fn new<'a>(
+            requires: Option<&Vec<Selection>>,
+            variable_usages: &[String],
+            data: &Value,
+            current_dir: &'a Path,
+            request: &Arc<Request>,
+            schema: &Arc<Schema>,
+        ) -> Result<Variables, FetchError> {
+            if let Some(requires) = requires {
+                let mut variables = Object::with_capacity(1 + variable_usages.len());
+                variables.extend(variable_usages.iter().filter_map(|key| {
+                    request.variables.as_ref().map(|v| {
+                        v.get(key)
+                            .map(|value| (key.clone(), value.clone()))
+                            .unwrap_or_default()
+                    })
+                }));
+
+                let mut paths = Vec::new();
+                let mut values = Vec::new();
+                data.select_values_and_paths(current_dir, |_path, value| {
+                    paths.push(_path);
+                    values.push(value)
+                })?;
+
+                let representations = Value::Array(
+                    values
+                        .into_iter()
+                        .flat_map(|value| match value {
+                            Value::Object(content) => {
+                                select_object(content, requires, schema).transpose()
+                            }
+                            _ => Some(Err(FetchError::ExecutionInvalidContent {
+                                reason: "not an object".to_string(),
+                            })),
+                        })
+                        .collect::<Result<Vec<_>, _>>()?,
                 );
-                let representations = selection::select(&response, current_dir, requires, &schema)?;
                 variables.insert("representations".into(), representations);
+
+                Ok(Variables { variables, paths })
+            } else {
+                Ok(Variables {
+                    variables: variable_usages
+                        .iter()
+                        .filter_map(|key| {
+                            request
+                                .variables
+                                .as_ref()
+                                .map(|v| v.get(key).map(|value| (key.clone(), value.clone())))
+                                .unwrap_or_default()
+                        })
+                        .collect::<Object>(),
+                    paths: Vec::new(),
+                })
             }
+        }
+    }
+
+    impl FetchNode {
+        pub(crate) async fn fetch_node<'a>(
+            &'a self,
+            data: &'a Value,
+            current_dir: &'a Path,
+            request: &'a Arc<Request>,
+            service_registry: Arc<dyn ServiceRegistry>,
+            schema: &'a Arc<Schema>,
```

Comment:
:yarn: Either you pass an `Arc<Something>` (because you need to clone), either you pass a `&Something`. But `&Arc<Something>` is unnecessary and restrict the way the caller can use this function.

```suggestion
            request: &'a Request,
            service_registry: Arc<dyn ServiceRegistry>,
            schema: &'a Schema,
```

Reply:
fixed in 634b640